File: rust/examples/todo-app/native-smokes/swift/LifecycleAppSmoke.swift
import Foundation

extension SyncularBoltClient: SyncularNativeJsonClient {}

private func expect(_ condition: Bool, _ message: String) {
    if !condition {
        fatalError(message)
    }
}

private func removeSqliteFiles(_ path: String) {
    let fileManager = FileManager.default
    for suffix in ["", "-wal", "-shm", "-journal"] {
        try? fileManager.removeItem(atPath: path + suffix)
    }
}

private func canonicalJson(_ object: Any) throws -> String {
    let data = try JSONSerialization.data(withJSONObject: object, options: [.sortedKeys])
    return String(data: data, encoding: .utf8)!
}

private func waitForEvent(
    from client: SyncularBoltClient,
    kind: String,
    commandId: String? = nil,
    timeoutMs: UInt64 = 5_000
) throws -> (SyncularNativeEvent, [String: Any]) {
    let deadline = Date().addingTimeInterval(Double(timeoutMs) / 1_000.0)
    var seen: [String] = []
    while Date() < deadline {
        if let eventJson = try client.nextEventJsonTimeout(timeoutMs: 50) {
            let event = try syncularDecodeNativeEvent(eventJson)
            let object = try JSONSerialization.jsonObject(with: Data(eventJson.utf8)) as! [String: Any]
            seen.append("\(event.kind):\(event.commandId ?? "-"):\(eventJson)")
            if event.kind == kind && (commandId == nil || event.commandId == commandId) {
                return (event, object)
            }
        }
    }
    fatalError("timed out waiting for native event \(kind) command \(commandId ?? "-"); seen \(seen.joined(separator: ", "))")
}

private func blobRef(fromPayload payload: [String: Any]) throws -> SyncularBlobRef {
    let data = try JSONSerialization.data(withJSONObject: payload, options: [.sortedKeys])
    return try JSONDecoder().decode(SyncularBlobRef.self, from: data)
}

private struct HostMaintenancePolicy {
    let isForeground: Bool
    let allowsExpensiveNetwork: Bool
    let allowsBackgroundWork: Bool
    let remainingBackgroundBudgetMs: UInt64

    var canProcessBlobUploads: Bool {
        allowsExpensiveNetwork && (isForeground || (allowsBackgroundWork && remainingBackgroundBudgetMs >= 2_000))
    }

    var canRunCompaction: Bool {
        isForeground || (allowsBackgroundWork && remainingBackgroundBudgetMs >= 1_000)
    }
}

@main
private enum LifecycleAppSmoke {
    static func main() throws {
        let outDir = CommandLine.arguments.dropFirst().first ?? NSTemporaryDirectory()
        try FileManager.default.createDirectory(atPath: outDir, withIntermediateDirectories: true)
        let dbPath = outDir + "/swift-lifecycle-app.sqlite"
        let blobPath = outDir + "/swift-lifecycle-blob.txt"
        removeSqliteFiles(dbPath)
        try Data("swift lifecycle blob".utf8).write(to: URL(fileURLWithPath: blobPath))

        let native = try SyncularBoltClient(open: SyncularBoltClientConfig(
            dbPath: dbPath,
            baseUrl: "http://127.0.0.1:9/sync",
            clientId: "swift-lifecycle-app",
            actorId: "user-rust",
            projectId: "project-rust",
            appSchemaJson: syncularNativeGeneratedAppSchemaJson,
            autoSyncLocalWrites: false
        ))
        defer { _ = try? native.shutdown() }
        expect(try native.startEventStream(capacity: 256), "Swift lifecycle should start native event stream")

        try assertSyncularNativeRuntimeManifestJson(try native.runtimeManifestJson())
        expect(try native.setAuthHeadersJson(headersJson: #"{"authorization":"Bearer lifecycle-swift"}"#), "Swift lifecycle app should accept auth headers")
        expect(try native.syncWorkerRunning(), "Swift lifecycle app should keep native worker hot")

        let query = TaskQuery
            .select()
            .filter(TaskQuery.userId.eq("user-rust"))
            .orderBy(TaskQuery.serverVersion.desc())
            .limit(20)
        let live = query.liveQuery(id: "swift-lifecycle-live", label: "Swift lifecycle")
        expect(try live.start(on: native).isEmpty, "Swift lifecycle live query should start empty")

        let blobCommandId = try native.enqueueStoreBlobFileJson(
            path: blobPath,
            optionsJson: #"{"mimeType":"text/plain"}"#
        )
        let (blobEvent, blobObject) = try waitForEvent(
            from: native,
            kind: "WorkerCommandCompleted",
            commandId: blobCommandId
        )
        expect(blobEvent.commandId == blobCommandId, "Swift lifecycle blob event should carry command id")
        let blobPayload = blobObject["payload_json"] as! [String: Any]
        let blobRef = try blobRef(fromPayload: blobPayload)
        let (blobStateEvent, _) = try waitForEvent(from: native, kind: "BlobUploadsChanged")
        expect(blobStateEvent.lifecycle?.blobUploads?.pending == 1, "Swift lifecycle should observe pending blob uploads")

        let rowId = "task-swift-lifecycle"
        let mutationCommandId = try native.queuedMutations.tasks.insert(NewTask(
            id: rowId,
            title: "",
            completed: 0,
            userId: "user-rust",
            projectId: "project-rust",
            image: blobRef
        ))
        let (mutationEvent, _) = try waitForEvent(
            from: native,
            kind: "LocalWriteCommitted",
            commandId: mutationCommandId
        )
        expect(mutationEvent.clientCommitId != nil, "Swift lifecycle mutation event should carry commit id")

        let crdtCommandId = try native.enqueueTaskTitleText(rowId: rowId, nextText: "Swift lifecycle title")
        let (crdtWriteEvent, _) = try waitForEvent(
            from: native,
            kind: "LocalWriteCommitted",
            commandId: crdtCommandId
        )
        expect(crdtWriteEvent.commandId == crdtCommandId, "Swift lifecycle CRDT write should carry command id")
        let (queryEvent, _) = try waitForEvent(from: native, kind: "QueriesChanged")
        _ = try waitForEvent(from: native, kind: "CrdtFieldChanged", commandId: crdtCommandId)

        let rows = try query.fetch(on: native)
        expect(rows.count == 1, "Swift lifecycle query should read queued row")
        expect(rows[0].id == rowId, "Swift lifecycle query should decode row id")
        expect(rows[0].title == "Swift lifecycle title", "Swift lifecycle query should decode CRDT title")
        expect(rows[0].image?.hash == blobRef.hash, "Swift lifecycle query should decode blob ref")
        expect(try live.refreshIfChanged(event: queryEvent, on: native)?.count == 1, "Swift lifecycle live query should refresh from native event")

        let syncCommandId = try native.resumeFromBackground()
        let (syncEvent, _) = try waitForEvent(
            from: native,
            kind: "SyncFailed",
            commandId: syncCommandId,
            timeoutMs: 8_000
        )
        expect(syncEvent.commandId == syncCommandId, "Swift lifecycle foreground resume sync failure should carry command id")

        let outbox = try native.outboxSummariesJson()
        expect(outbox.contains(mutationEvent.clientCommitId!), "Swift lifecycle outbox should contain queued mutation")
        expect(outbox.contains(crdtWriteEvent.clientCommitId!), "Swift lifecycle outbox should contain queued CRDT write")

        let restrictedBackground = HostMaintenancePolicy(
            isForeground: false,
            allowsExpensiveNetwork: false,
            allowsBackgroundWork: false,
            remainingBackgroundBudgetMs: 0
        )
        expect(!restrictedBackground.canProcessBlobUploads, "Swift restricted background policy should not process blob uploads")
        expect(!restrictedBackground.canRunCompaction, "Swift restricted background policy should not run compaction")

        let foregroundPolicy = HostMaintenancePolicy(
            isForeground: true,
            allowsExpensiveNetwork: true,
            allowsBackgroundWork: false,
            remainingBackgroundBudgetMs: 0
        )
        if foregroundPolicy.canProcessBlobUploads {
            let uploadCommandId = try native.enqueueProcessBlobUploadQueue()
            let (uploadEvent, _) = try waitForEvent(
                from: native,
                kind: "WorkerCommandCompleted",
                commandId: uploadCommandId,
                timeoutMs: 8_000
            )
            expect(uploadEvent.commandId == uploadCommandId, "Swift lifecycle queued blob upload processing should carry command id")
        }
        if foregroundPolicy.canRunCompaction {
            let compactCommandId = try native.enqueueCompactStorageJson(optionsJson: #"{"olderThanMs":0}"#)
            let (compactEvent, _) = try waitForEvent(
                from: native,
                kind: "WorkerCommandCompleted",
                commandId: compactCommandId
            )
            expect(compactEvent.commandId == compactCommandId, "Swift lifecycle queued compaction should carry command id")
        }

        expect(try live.stop(on: native), "Swift lifecycle live query should unregister")
        expect(try native.shutdown(), "Swift lifecycle app should shut down native client")

        _ = try canonicalJson(blobPayload)
        print("Swift lifecycle app smoke passed")
    }
}

File: rust/examples/todo-app/native-smokes/kotlin/LifecycleAppSmoke.kt
import dev.syncular.client.SyncularBoltClient
import dev.syncular.client.SyncularBoltClientConfig
import java.io.File
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.long

private class BoltNativeClient(
    private val client: SyncularBoltClient,
) : SyncularNativeJsonClient {
    override fun applyMutationJson(mutationJson: String, localRowJson: String?): String =
        client.applyMutationJson(mutationJson, localRowJson)

    override fun enqueueMutationJson(mutationJson: String, localRowJson: String?): String =
        client.enqueueMutationJson(mutationJson, localRowJson)

    override fun openCrdtFieldJson(requestJson: String): String =
        client.openCrdtFieldJson(requestJson)

    override fun applyCrdtFieldTextJson(requestJson: String): String =
        client.applyCrdtFieldTextJson(requestJson)

    override fun applyCrdtFieldYjsUpdateJson(requestJson: String): String =
        client.applyCrdtFieldYjsUpdateJson(requestJson)

    override fun enqueueCrdtFieldYjsUpdateJson(requestJson: String): String =
        client.enqueueCrdtFieldYjsUpdateJson(requestJson)

    override fun enqueueCrdtFieldTextJson(requestJson: String): String =
        client.enqueueCrdtFieldTextJson(requestJson)

    override fun enqueueCrdtFieldCompactionJson(requestJson: String): String =
        client.enqueueCrdtFieldCompactionJson(requestJson)

    override fun materializeCrdtFieldJson(requestJson: String): String =
        client.materializeCrdtFieldJson(requestJson)

    override fun snapshotCrdtFieldStateVectorJson(requestJson: String): String =
        client.snapshotCrdtFieldStateVectorJson(requestJson)

    override fun compactCrdtFieldJson(requestJson: String): String =
        client.compactCrdtFieldJson(requestJson)

    override fun queryJson(requestJson: String): String =
        client.queryJson(requestJson)

    override fun registerQueryJson(queryJson: String): String =
        client.registerQueryJson(queryJson)

    override fun unregisterQuery(id: String): Boolean =
        client.unregisterQuery(id)
}

private data class NativeEventEnvelope(
    val event: SyncularNativeEvent,
    val json: JsonObject,
)

private fun expect(condition: Boolean, message: String) {
    if (!condition) error(message)
}

private val lifecycleJson = Json { ignoreUnknownKeys = true }

private fun removeSqliteFiles(path: String) {
    listOf("", "-wal", "-shm", "-journal").forEach { suffix ->
        File(path + suffix).delete()
    }
}

private fun waitForEvent(
    client: SyncularBoltClient,
    kind: String,
    commandId: String? = null,
    timeoutMs: Long = 5_000,
): NativeEventEnvelope {
    val deadline = System.currentTimeMillis() + timeoutMs
    val seen = mutableListOf<String>()
    while (System.currentTimeMillis() < deadline) {
        val eventJson = client.pollEventJsonTimeout(100uL) ?: continue
        val event = syncularDecodeNativeEvent(eventJson)
        val json = lifecycleJson.parseToJsonElement(eventJson).jsonObject
        seen += "${event.kind}:${event.commandId ?: "-"}"
        if (event.kind == kind && (commandId == null || event.commandId == commandId)) {
            return NativeEventEnvelope(event, json)
        }
    }
    error("timed out waiting for native event $kind command ${commandId ?: "-"}; seen ${seen.joinToString()}")
}

private fun JsonObject.str(key: String): String =
    this[key]?.jsonPrimitive?.content ?: error("missing string field $key")

private fun JsonObject.longValue(key: String): Long =
    this[key]?.jsonPrimitive?.long ?: error("missing integer field $key")

fun main(args: Array<String>) {
    val outDir = File(args.firstOrNull() ?: System.getProperty("java.io.tmpdir"))
    outDir.mkdirs()
    val dbPath = File(outDir, "kotlin-lifecycle-app.sqlite").absolutePath
    val blobPath = File(outDir, "kotlin-lifecycle-blob.txt")
    removeSqliteFiles(dbPath)
    blobPath.writeText("kotlin lifecycle blob")

    val raw = SyncularBoltClient(
        SyncularBoltClientConfig(
            dbPath = dbPath,
            baseUrl = "http://127.0.0.1:9/sync",
            clientId = "kotlin-lifecycle-app",
            actorId = "user-rust",
            projectId = "project-rust",
            appSchemaJson = syncularNativeGeneratedAppSchemaJson,
            autoSyncLocalWrites = false,
        ),
    )
    val client = BoltNativeClient(raw)
    try {
        val manifest = raw.runtimeManifestJson()
        expect(manifest.contains("\"storage_backend\":\"diesel-sqlite\""), "Kotlin lifecycle app should expose diesel sqlite manifest")
        expect(raw.setAuthHeadersJson("""{"authorization":"Bearer lifecycle-kotlin"}"""), "Kotlin lifecycle app should accept auth headers")
        expect(raw.syncWorkerRunning(), "Kotlin lifecycle app should keep native worker hot")

        val query = TaskQuery
            .select()
            .filter(TaskQuery.userId.eq("user-rust"))
            .orderBy(TaskQuery.serverVersion.desc())
            .limit(20)
        val live = query.liveQuery(id = "kotlin-lifecycle-live", label = "Kotlin lifecycle")
        expect(live.start(client).isEmpty(), "Kotlin lifecycle live query should start empty")

        val blobCommandId = raw.enqueueStoreBlobFileJson(
            blobPath.absolutePath,
            """{"mimeType":"text/plain"}""",
        )
        val blobEnvelope = waitForEvent(raw, kind = "WorkerCommandCompleted", commandId = blobCommandId)
        expect(blobEnvelope.event.commandId == blobCommandId, "Kotlin lifecycle blob event should carry command id")
        val blobPayload = blobEnvelope.json["payload_json"]!!.jsonObject
        val blobRef = SyncularBlobRef(
            hash = blobPayload.str("hash"),
            size = blobPayload.longValue("size"),
            mimeType = blobPayload.str("mimeType"),
        )

        val rowId = "task-kotlin-lifecycle"
        val mutationCommandId = client.enqueueNewTask(
            NewTask(
                id = rowId,
                title = "",
                completed = 0,
                userId = "user-rust",
                projectId = "project-rust",
                image = blobRef,
            ),
        )
        val mutationEvent = waitForEvent(raw, kind = "LocalWriteCommitted", commandId = mutationCommandId).event
        expect(mutationEvent.clientCommitId != null, "Kotlin lifecycle mutation event should carry commit id")

        val crdtCommandId = client.enqueueTaskTitleText(rowId = rowId, nextText = "Kotlin lifecycle title")
        val crdtWriteEvent = waitForEvent(raw, kind = "LocalWriteCommitted", commandId = crdtCommandId).event
        expect(crdtWriteEvent.commandId == crdtCommandId, "Kotlin lifecycle CRDT write should carry command id")
        val queryEvent = waitForEvent(raw, kind = "QueriesChanged").event
        waitForEvent(raw, kind = "CrdtFieldChanged", commandId = crdtCommandId)

        val rows = query.fetch(client)
        expect(rows.size == 1, "Kotlin lifecycle query should read queued row")
        expect(rows[0].id == rowId, "Kotlin lifecycle query should decode row id")
        expect(rows[0].title == "Kotlin lifecycle title", "Kotlin lifecycle query should decode CRDT title")
        expect(rows[0].image?.hash == blobRef.hash, "Kotlin lifecycle query should decode blob ref")
        expect(live.refreshIfChanged(queryEvent, client)?.size == 1, "Kotlin lifecycle live query should refresh from native event")

        val syncCommandId = raw.enqueueSyncNow()
        val syncEvent = waitForEvent(
            raw,
            kind = "SyncFailed",
            commandId = syncCommandId,
            timeoutMs = 8_000,
        ).event
        expect(syncEvent.commandId == syncCommandId, "Kotlin lifecycle sync failure should carry command id")

        val outbox = raw.outboxSummariesJson()
        expect(outbox.contains(mutationEvent.clientCommitId!!), "Kotlin lifecycle outbox should contain queued mutation")
        expect(outbox.contains(crdtWriteEvent.clientCommitId!!), "Kotlin lifecycle outbox should contain queued CRDT write")
        expect(live.stop(client), "Kotlin lifecycle live query should unregister")
        expect(raw.shutdown(), "Kotlin lifecycle app should shut down native client")

        println("Kotlin lifecycle app smoke passed")
    } finally {
        runCatching { raw.shutdown() }
        raw.close()
    }
}
